### PR TITLE
Load next config

### PR DESCRIPTION
**Pull request by Brooooooklyn**
Load next.config.js via turbopack-node

Closes WEB-158

### CODE REVIEW

File: crates/turbopack-node/src/read_config.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            };
+            emit(bootstrap.cell().into(), self.intermediate_output_path()).await?;
+            let mut operation = pool.operation().await?;
+            let output = eval_js_operation(
+                &mut operation,
+                EvalJavaScriptOutgoingMessage::LoadNextConfig {
+                    path: to_sys_path(self.path())
+                        .await?
+                        .and_then(|p| p.to_str().map(|s| s.to_string()))
+                        .ok_or_else(|| anyhow!("Invalid next.config.js path"))?,
```
Comment (ForsakenHarmony) [inline crates/turbopack-node/src/read_config.rs:51]:
```suggestion
                        .ok_or_else(|| anyhow!("Invalid config path"))?,
```

this crate should not reference anything next specific

File: crates/turbopack-node/src/lib.rs [hunk 2]
```diff
@@ -623,6 +646,19 @@ async fn proxy_error(
...
+) -> Result<Vec<u8>> {
+    operation.send(content).await?;
+    match operation.recv().await? {
+        EvalJavaScriptIncomingMessage::Error(err) => {
+            bail!(err.print(Default::default(), None).await?);
+        }
+        EvalJavaScriptIncomingMessage::JavaScriptValue { data } => Ok(data),
+    }
+}
+
```
Comment (ForsakenHarmony) [inline crates/turbopack-node/src/lib.rs:661]: maybe we can move this to its own file

File: crates/next-core/js/src/read-config/next.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+};
+
+type IpcOutgoingMessage = {
+  type: "javaScriptValue";
+  data: number[];
+};
+
+const ipc = IPC as Ipc<IpcIncomingMessage, IpcOutgoingMessage>;
+
+const CACHED_CONFIGS = new Map<string, any>();
```
Comment (ForsakenHarmony) [inline crates/next-core/js/src/read-config/next.ts:17]: I think we're better off caching this on the rust side with turbo tasks?

Also this cache is never used

File: crates/next-core/js/src/read-config/next.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      case "loadNextConfig": {
+        const nextConfig = await loadConfig(PHASE_DEVELOPMENT_SERVER, msg.path);
+        CACHED_CONFIGS.set(msg.path, nextConfig);
+        // @ts-expect-error
+        nextConfig.rewrites = await nextConfig.rewrites?.();
+        // @ts-expect-error
+        nextConfig.redirects = await nextConfig.redirects?.();
+        await ipc.send({
+          type: "javaScriptValue",
+          data: Array.from(Buffer.from(JSON.stringify(nextConfig))),
```
Comment (ForsakenHarmony) [inline crates/next-core/js/src/read-config/next.ts:33]:
```suggestion
          data: JSON.stringify(nextConfig)),
```

why not just

File: crates/next-core/js/src/read-config/next.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import IPC, { Ipc } from "@vercel/turbopack-next/internal/ipc";
```
Comment (ForsakenHarmony) [inline crates/next-core/js/src/read-config/next.ts:1]: maybe move this to `src/entry/config/next.ts`

File: crates/next-core/src/next_config.rs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+use turbopack_ecmascript::{
+    EcmascriptInputTransform, EcmascriptInputTransformsVc, EcmascriptModuleAssetType,
+    EcmascriptModuleAssetVc,
+};
+use turbopack_node::read_config::{JavaScriptConfig, JavaScriptConfigVc, JavaScriptValue};
+
+use crate::embed_js::next_js_file;
+
+#[turbo_tasks::value(transparent)]
+pub struct NextConfigValue(NextConfig);
```
Comment (ForsakenHarmony) [inline crates/next-core/src/next_config.rs:19]: Why is this a separate struct?

File: crates/next-core/js/src/read-config/next.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import loadConfig from "next/dist/server/config";
+import { PHASE_DEVELOPMENT_SERVER } from "next/dist/shared/lib/constants";
+
+type IpcIncomingMessage = {
+  type: "loadNextConfig" | "loadPostcssConfig";
```
Comment (ForsakenHarmony) [inline crates/next-core/js/src/read-config/next.ts:6]:
```suggestion
  type: "loadNextConfig";
```

File: crates/turbopack-node/src/read_config.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+pub enum JavaScriptValue {
+    Value(Rope),
+    Stream(#[turbo_tasks(trace_ignore)] Vec<u8>),
+}
+
+#[turbo_tasks::value(transparent)]
+pub struct JavaScriptValueArray(Vec<JavaScriptValueVc>);
+
+#[turbo_tasks::value_trait]
+pub trait JavaScriptConfig {
```
Comment (ForsakenHarmony) [inline crates/turbopack-node/src/read_config.rs:24]: This could just be a struct (or even just a function)?

File: crates/turbopack-node/src/read_config.rs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+use crate::{
+    bootstrap::NodeJsBootstrapAsset, emit, eval_js_operation, pool::NodeJsPool,
+    EvalJavaScriptOutgoingMessage,
+};
+
+#[turbo_tasks::value(shared)]
+#[derive(Clone)]
+pub enum JavaScriptValue {
+    Value(Rope),
+    Stream(#[turbo_tasks(trace_ignore)] Vec<u8>),
```
Comment (ForsakenHarmony) [inline crates/turbopack-node/src/read_config.rs:17]: Do we really need this?

File: crates/turbopack-node/src/read_config.rs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                        .await?
+                        .and_then(|p| p.to_str().map(|s| s.to_string()))
+                        .ok_or_else(|| anyhow!("Invalid next.config.js path"))?,
+                },
+            )
+            .await?;
+            Ok(JavaScriptValue::Value(output.into()).cell())
+        } else {
+            Err(anyhow!("can only render from a disk filesystem"))
+        }
```
Comment (ForsakenHarmony) [inline crates/turbopack-node/src/read_config.rs:58]: exit early instead of wrapping and indenting

File: crates/turbopack-node/src/read_config.rs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+#[turbo_tasks::value(shared)]
+#[derive(Clone)]
+pub enum JavaScriptValue {
+    Value(Rope),
+    Stream(#[turbo_tasks(trace_ignore)] Vec<u8>),
+}
+
+#[turbo_tasks::value(transparent)]
+pub struct JavaScriptValueArray(Vec<JavaScriptValueVc>);
```
Comment (ForsakenHarmony) [inline crates/turbopack-node/src/read_config.rs:21]: this is unused?

File: crates/turbopack-node/src/read_config.rs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+            to_sys_path(self.intermediate_output_path().join("read-config.js")).await?,
+        ) {
+            let pool = NodeJsPool::new(cwd, entrypoint, HashMap::new(), 1);
+            let bootstrap = NodeJsBootstrapAsset {
+                path: self.intermediate_output_path().join("read-config.js"),
+                chunk_group: ChunkGroupVc::from_chunk(
+                    self.entry()
+                        .as_evaluated_chunk(self.chunking_context(), None),
+                ),
+            };
```
Comment (ForsakenHarmony) [inline crates/turbopack-node/src/read_config.rs:42]: could probably use our `get_intermediate_asset` fn here

File: crates/turbopack-node/src/read_config.rs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    fn entry(&self) -> EcmascriptModuleAssetVc;
+    fn chunking_context(&self) -> ChunkingContextVc;
+    fn intermediate_output_path(&self) -> FileSystemPathVc;
+
+    async fn load(&self) -> Result<JavaScriptValueVc> {
+        if let (Some(cwd), Some(entrypoint)) = (
+            to_sys_path(self.intermediate_output_path()).await?,
+            to_sys_path(self.intermediate_output_path().join("read-config.js")).await?,
+        ) {
+            let pool = NodeJsPool::new(cwd, entrypoint, HashMap::new(), 1);
```
Comment (ForsakenHarmony) [inline crates/turbopack-node/src/read_config.rs:35]: I feel like we should create the pool after we emit the asset

File: crates/turbopack-node/src/lib.rs [hunk 14]
```diff
@@ -480,13 +444,72 @@ enum RenderProxyIncomingMessage {
     Error(StructuredError),
 }
 
+#[derive(Serialize)]
+#[serde(tag = "type", rename_all = "camelCase")]
+enum EvalJavaScriptOutgoingMessage {
+    LoadNextConfig { path: String },
```
Comment (ForsakenHarmony) [inline crates/turbopack-node/src/lib.rs:450]:
```suggestion
    LoadConfig { path: String },
```

File: crates/next-core/src/next_config.rs [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+                EcmascriptInputTransformsVc::cell(vec![
+                    EcmascriptInputTransform::React { refresh: false },
+                    EcmascriptInputTransform::TypeScript,
+                ]),
+                context.environment(),
+            ),
+            chunking_context,
+            intermediate_output_path: output_root,
+        }
+        .cell()
```
Comment (ForsakenHarmony) [inline crates/next-core/src/next_config.rs:118]:
```suggestion
        let entry_module = EcmascriptModuleAssetVc::new(
            VirtualAssetVc::new(
                output_root.join("next.js"),
                next_js_file("read-config/next.ts").into(),
            )
            .into(),
            context,
            Value::new(EcmascriptModuleAssetType::Typescript),
            EcmascriptInputTransformsVc::cell(vec![
                EcmascriptInputTransform::React { refresh: false },
                EcmascriptInputTransform::TypeScript,
            ]),
            context.environment(),
        );

        NextConfigLoader {
            path: project_root,
            entry_module,
            chunking_context,
            intermediate_output_path: output_root,
        }
        .cell()
```

File: crates/next-core/src/next_config.rs [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+#[turbo_tasks::value_impl]
+impl NextConfigLoaderVc {
+    #[turbo_tasks::function]
+    pub fn new(
+        context: AssetContextVc,
+        chunking_context: ChunkingContextVc,
+        project_root: FileSystemPathVc,
+        output_root: FileSystemPathVc,
```
Comment (ForsakenHarmony) [inline crates/next-core/src/next_config.rs:97]:
```suggestion
        intermediate_output_path: FileSystemPathVc,
```

**Review by ForsakenHarmony — CHANGES_REQUESTED** [review]

**Comment by alexkirsz** [comment]
Benchmarks look pretty rough: https://github.com/vercel/turbo/pull/2955#issuecomment-1342887336

File: crates/turbopack-node/src/read_config.rs (hunk 12 above)
Comment (Brooooooklyn) [reply crates/turbopack-node/src/read_config.rs:42]: dose calling `get_intermediate_asset` will overwrite the entry file?

File: crates/next-core/src/next_config.rs (hunk 6 above)
Comment (Brooooooklyn) [reply crates/next-core/src/next_config.rs:19]: I remember use `#[turbo_tasks::value]` directly will implement different`Serialize/Deserialize` trait for the `NextConfig`

File: crates/turbopack-node/src/lib.rs (hunk 14 above)
Comment (Brooooooklyn) [reply crates/turbopack-node/src/lib.rs:450]: let's keep the rpc name as `LoadNextConfig`,  the follow-up will add more different types of config reading logic

**Comment by Brooooooklyn** [comment]
> Benchmarks look pretty rough

@alexkirsz fixed

File: crates/turbopack-node/src/lib.rs (hunk 14 above)
Comment (Brooooooklyn) [reply crates/turbopack-node/src/lib.rs]: resolved

File: crates/next-core/src/next_client/context.rs [hunk 17]
```diff
@@ -238,7 +238,7 @@ pub async fn get_client_runtime_entries(
             .as_request();
 
     let mut runtime_entries = vec![RuntimeEntry::Ecmascript(
-        ProcessEnvAssetVc::new(project_root, env_for_js(env, true)).into(),
+        ProcessEnvAssetVc::new(project_root, env_for_js(env, true, None)).into(),
```
Comment (jridgewell) [inline crates/next-core/src/next_client/context.rs:241]: Why isn't a next config used here?

File: crates/next-core/src/next_config.rs (hunk 6 above)
Comment (jridgewell) [reply crates/next-core/src/next_config.rs:19]: You can use `#[turbo_tasks::value(serialization = "custom")]` to manually implement serialization/deserialization.

File: crates/next-core/src/next_config.rs [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+    #[serde(rename(deserialize = "dangerouslyAllowSVG"))]
+    pub dangerously_allow_svg: bool,
+    pub content_security_policy: String,
+    pub remote_patterns: Vec<RemotePattern>,
+    pub unoptimized: bool,
+}
+
+impl Default for ImageConfig {
+    fn default() -> Self {
+        Self {
```
Comment (jridgewell) [inline crates/next-core/src/next_config.rs:62]: Nit: I should have done this myself when adding the old default, but let's add a link:
```suggestion
        // https://github.com/vercel/next.js/blob/327634eb/packages/next/shared/lib/image-config.ts#L100-L114
        Self {
```

File: crates/next-core/src/next_config.rs [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+        let config_chunk = EcmascriptModuleAssetVc::new(
+            config_asset.into(),
+            context,
+            Value::new(EcmascriptModuleAssetType::Ecmascript),
+            EcmascriptInputTransformsVc::cell(vec![]),
+            context.environment(),
+        )
+        .as_ecmascript_chunk_placeable();
+        chunks = Some(EcmascriptChunkPlaceablesVc::cell(vec![config_chunk]));
+    }
```
Comment (jridgewell) [inline crates/next-core/src/next_config.rs:188]: Nit: having an `if … = if {…}` was a bit hard to parse. Can we split this into 2 statements?

File: crates/next-core/src/next_config.rs [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+) -> Result<NextConfigValueVc> {
+    let mut chunks = None;
+    let next_config_mjs_path = project_root.join("next.config.mjs");
+    let next_config_js_path = project_root.join("next.config.js");
+    if let Some(config_asset) = if matches!(
+        &*next_config_mjs_path.read().await?,
+        FileContent::Content(_)
+    ) {
+        Some(SourceAssetVc::new(next_config_mjs_path))
+    } else if matches!(&*next_config_js_path.read().await?, FileContent::Content(_)) {
```
Comment (jridgewell) [inline crates/next-core/src/next_config.rs:174]: Nit: using `path.get_type()` will allow us to see if the file exists without actually reading the file content. +1 laziness.

File: crates/turbopack-node/js/src/evaluate.ts [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+        const { execute } = eval("require")(msg.filepath);
+        if (typeof execute !== "function") {
+          console.error(
+            `Expected ${msg.filepath} to export a function named "execute"`
+          );
+          process.exit(1);
+        }
+        const value = await execute.apply(null, msg.arguments);
+        await ipc.send({
+          type: "javaScriptValue",
```
Comment (jridgewell) [inline crates/turbopack-node/js/src/evaluate.ts:31]: Nit: Can we fix this weird casing?

**Review by jridgewell — APPROVED** [review]

File: crates/next-core/src/app_source.rs [hunk 22]
```diff
@@ -222,10 +223,10 @@ pub async fn create_app_source(
     browserslist_query: &str,
     externals: StringsVc,
 ) -> Result<ContentSourceVc> {
-    let project_root = wrap_with_next_js_fs(project_root);
+    let project_path = wrap_with_next_js_fs(project_root);
```
[the remark this answers is not in the document]
Comment (Brooooooklyn) [reply crates/next-core/src/app_source.rs:226]: `project_root` here is shadow the parameter `project_root`, and the parameter needs to use below

File: crates/next-core/src/server_rendered_source.rs [hunk 23]
```diff
@@ -120,8 +122,24 @@ pub async fn create_server_rendered_source(
...
+    let next_config_value = load_next_config(
+        context,
+        DevChunkingContextVc::builder(
+            project_path,
+            output_path,
+            output_path.join("chunks"),
+            get_client_assets_path(server_root, Value::new(ContextType::Pages { pages_dir })),
+        )
+        .build(),
+        project_root,
```
[the remark this answers is not in the document]
Comment (Brooooooklyn) [reply crates/next-core/src/server_rendered_source.rs:134]: `project_path` is wrapped with `next_fs` and it's the `AttachedFileSystemVc` inner it. I need `DiskFileSystemVc` type `FileSystemPathVc` here

File: crates/next-core/src/next_client/context.rs (hunk 17 above)
Comment (Brooooooklyn) [reply crates/next-core/src/next_client/context.rs:242]: Pass `NextConfigValueVc` into `env_for_js` here need to pass `NextConfigValueVc` into `fn app_context` and the `NextConfigVaueVc` is built from it.
It's a temporary solution, I'll build a separate `AssetContext` for `fn load_next_config` later

File: crates/next-core/src/next_import_map.rs [hunk 24]
```diff
@@ -66,6 +67,25 @@ pub fn get_next_client_import_map(
...
+    let package_root = attached_next_js_package_path(project_path);
+
+    insert_package_alias(
+        &mut import_map,
+        &format!("{VIRTUAL_PACKAGE_NAME}/"),
+        package_root,
+    );
+
+    import_map.insert_exact_alias("next", ImportMapping::External(None).into());
+    import_map.insert_wildcard_alias("next/", ImportMapping::External(None).into());
```
Comment (jridgewell) [inline crates/next-core/src/next_import_map.rs:84]: Does this mean that `npm install next` after startup won't rebuild the config?

File: crates/next-dev/src/lib.rs [hunk 25]
```diff
@@ -315,14 +306,15 @@ async fn source(
         dev_server_root,
         env,
         &browserslist_query,
+        next_config,
```
Comment (jridgewell) [inline crates/next-dev/src/lib.rs:309]: Much nicer to pass this down.

File: crates/next-dev/src/lib.rs [hunk 26]
```diff
@@ -286,8 +275,10 @@ async fn source(
     let project_path = fs.root().join(project_relative);
 
     let env = load_env(project_path);
+    let config_output_root = output_fs.root().join(".next/config");
+    let next_config = load_next_config(project_path, config_output_root);
```
Comment (jridgewell) [inline crates/next-dev/src/lib.rs:279]: I think the `web_source` should also be using this, so that we can load the appropriate image config for their external references. (I can't comment on the appropriate line below).

File: crates/next-core/src/next_config.rs [hunk 27]
```diff
@@ -115,6 +126,7 @@ pub enum RemotePatternProtocal {
 #[serde(rename_all = "camelCase")]
 pub struct ExperimentalConfig {
     pub server_components_external_packages: Option<Vec<String>>,
+    pub app_dir: Option<bool>,
```
Comment (jridgewell) [inline crates/next-core/src/next_config.rs:129]: Wooo!

**Review by jridgewell — APPROVED** [review]

File: crates/next-core/src/next_import_map.rs (hunk 24 above)
Comment (sokra) [reply crates/next-core/src/next_import_map.rs:84]: Yeah. Not perfect, but it's for performance reasons. We also do that in the SSR code. Next pulls in so much stuff, which would add 2 seconds to the bundling and emitting. Mostly because it pulls in all the build related stuff too like webpack, loaders, plugins and the whole caniuse db...

We should probably optimize that in next.js to not depend on everything, but I guess it's fine for now.
When updating next.js you get a new turbopack binary anyway, which requires a restart.

File: crates/next-dev/src/lib.rs (hunk 26 above)
Comment (sokra) [reply crates/next-dev/src/lib.rs:279]: Passed it to it too. Not sure if someone uses next/image with CSR mode (non-next), but at least it's consistent now.

File: crates/next-core/src/next_config.rs (hunk 27 above)
Comment (sokra) [reply crates/next-core/src/next_config.rs:130]: Little change, but it actually broken the benchmarks since we need a next.config.js for the benchmarks now...

File: crates/turbopack-node/src/source_map/trace.rs [hunk 28]
```diff
@@ -20,7 +22,23 @@ pub struct StackFrame {
...
 
+impl Display for StackFrame {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.get_pos() {
+            Some((l, c)) => match &self.name {
+                Some(n) => write!(f, "{} ({}:{}:{})", n, self.file, l, c),
+                None => write!(f, "{}:{}:{}", self.file, l, c),
+            },
+            None => write!(f, "{}", self.file),
+        }
```
Comment (jridgewell) [inline crates/turbopack-node/src/source_map/trace.rs:33]: Nit:
```suggestion
        self.with_path(&self.file).fmt(f)
```

File: crates/turbopack-node/src/source_map/trace.rs [hunk 29]
```diff
@@ -20,7 +22,23 @@ pub struct StackFrame {
...
+}
+
 impl StackFrame {
+    pub fn with_path<'a>(&'a self, path: &'a str) -> StackFrameWithPath<'a> {
+        StackFrameWithPath { frame: self, path }
+    }
+
     pub fn get_pos(&self) -> Option<(usize, usize)> {
         match (self.line, self.column) {
             (Some(l), Some(c)) => Some((l, c)),
```
Comment (jridgewell) [inline crates/turbopack-node/src/source_map/trace.rs:44]: Nit: I recently learned about [Option::zip](https://doc.rust-lang.org/std/option/enum.Option.html#method.zip): `self.line.zip(self.column)`

File: crates/turbopack-dev-server/src/html.rs [hunk 30]
```diff
@@ -187,7 +187,7 @@ impl VersionedContent for DevHtmlAssetContent {
     async fn update(self_vc: DevHtmlAssetContentVc, from_version: VersionVc) -> Result<UpdateVc> {
         let from_version = DevHtmlAssetVersionVc::resolve_from(from_version)
             .await?
-            .expect("version must be an `DevHtmlAssetVersionVc`");
+            .ok_or_else(|| anyhow!("version must be an `DevHtmlAssetVersionVc`"))?;
```
Comment (jridgewell) [inline crates/turbopack-dev-server/src/html.rs:190]: Nit: Maybe just `.context("…")` since it's a static string anyways?

File: crates/next-core/js/src/ipc/index.ts [hunk 31]
```diff
@@ -0,0 +1,144 @@
+import { createConnection } from "node:net";
```
Comment (jridgewell) [inline crates/next-core/js/src/ipc/index.ts:1]: The duplication between `next-core` and `turbopack-node` is really confusing. Do we need both?

File: crates/next-core/src/next_client/context.rs [hunk 32]
```diff
@@ -238,7 +240,8 @@ pub async fn get_client_runtime_entries(
             .as_request();
 
     let mut runtime_entries = vec![RuntimeEntry::Ecmascript(
-        ProcessEnvAssetVc::new(project_root, env_for_js(env, true)).into(),
+        //TODO
```
Comment (jridgewell) [inline crates/next-core/src/next_client/context.rs:243]: What's TODO?

File: crates/next-core/src/next_config.rs [hunk 33]
```diff
@@ -192,6 +193,17 @@ impl NextConfigVc {
                 .unwrap_or_default(),
         ))
     }
+
+    #[turbo_tasks::function]
+    pub async fn image_config(self) -> Result<ImageConfigVc> {
+        let mut image_config = self.await?.images.clone();
+        // TODO is this the right place to do this?
+        image_config.remote_patterns.push(RemotePattern {
```
Comment (jridgewell) [inline crates/next-core/src/next_config.rs:201]: I think we can drop this entirely now. It was just done because we couldn't load the next config, and we need to allow everything so that the things that were supposed to be allowed in the config would work. But now, we can load the config and know what was actually supposed to be allowed. If they didn't allow a host, then it should fail.

The reason it was still present in earlier PR revisions is because the web_source didn't load the config, but now that works, too.

**Review by jridgewell — COMMENTED** [review]
GH didn't show me the full diff of new changes…

File: crates/next-core/js/src/ipc/index.ts (hunk 31 above)
Comment (sokra) [reply crates/next-core/js/src/ipc/index.ts:1]: Currently both are used, but that's not great. Eventually next-core should use turbopack-node for the SSR stuff too. But let's move that to a later PR

File: crates/turbopack-dev-server/src/html.rs (hunk 30 above)
Comment (sokra) [reply crates/turbopack-dev-server/src/html.rs:190]: Does context() work on Option?

File: crates/next-core/src/next_config.rs (hunk 33 above)
Comment (sokra) [reply crates/next-core/src/next_config.rs:201]: Sounds good. Follow-up PR

File: crates/next-core/src/next_client/context.rs (hunk 32 above)
Comment (sokra) [reply crates/next-core/src/next_client/context.rs:243]: I guess that's solved by passing the next_config here...? Maybe @Brooooooklyn knows more.

**Review by sokra — APPROVED** [review]
I'm merging this even while there are some unsolved nitpicks. These can be fixed in follow-up PRs.
Merging this should unblock some follow-up work...
